text/0000-prelude-2021.md

```diff
@@ -0,0 +1,164 @@
+
+Especially when coming from another language where hashmaps and sets are language built-ins with literals provided as part of the language, it can be a little surprising at first that these have to be imported before use. Their usage can be especially common in script-like code, for which effortless usage will be especially helpful.
+
+While a number of third-party hashmaps are in common use (most of which reexport or wrap the std `HashMap` with a different hasher), bringing the std `HashMap` into the prelude should not cause issues, since explicit imports will always override prelude imports.
+
+## `std::{{fs::File}, path::{Path, PathBuf}}`
+
+Are very commonly used, not least of which in script-like CLI tools.
+
+## `std::sync::{Arc, Mutex, RwLock}`
+
+Help write threaded code. A potential concern is that async code may want to use async variants of these, although even then the [sync Mutex](https://docs.rs/tokio/1.1.1/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use) may provide better performance.
+
+Use of alternative implementations (in particular from parking_lot) is fairly common.
+
+## `std::borrow::Cow`
+
+Unlikely to clash with other types, yet this type can be of great help in bridging the gap between references and owned types, particular for `String`/`str` and `Vec`/`slice` types.
+
+## `std::cmp::{min, max}`
+
+It has come up in [discussions](https://internals.rust-lang.org/t/am-i-the-only-one-confused-by-a-min-b-and-a-max-b/13252) that the method forms of `min()` and `max()` attached to std numeric types can be confusing to understand. In order to level the playing field with the two-argument free functions of the same names, it would be helpful to import these into the prelude.
+
+## Other traits that have been suggested for inclusion
+
+Other traits that have been suggested as prelude candidates:
+
+- `std::ops::Not`: for chaining purposes, it is sometimes useful to have an trailing `.not()` call rather than the prefix `!` operator. Therefore, [it has been suggested](https://internals.rust-lang.org/t/pre-rfc-adding-bool-not-method/13935) that the `Not` trait which brings that method in scope could be added to the prelude. This RFC author feels this use case is better served by adding an inherent impl to `bool`, since that serves the majority of the same use case with less complexity.
+- `std::fmt::Display`: users of this trait generally don't need to import it because the `ToString` trait which relies on the `Display` implementation is already in the prelude. Implementers of the `Display` trait however need several other items from `std::fmt` to do so; therefore, just importing `Display` into the prelude does not help much. This RFC suggests adding `fmt` to the prelude to improve this use case.
+- `std::fmt::Debug`: similar to `Display`, although there's no `to_debug()`. However, usage will usually go through `dbg!()` or the formatting mechanism (as a `{:?}` format string).
```

Comment:
Caveat: The `fmt` method will collide where the user originally only imported one of the traits.

E.g:

```rust
use core::fmt;
use core::fmt::Display;

struct MyStruct(u32);

impl Display for MyStruct {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f) // Just delegate to the inner impl
    }
}
```

Reply:
I'm not sure I understand what you're getting at. That code seems to compile just fine?

Reply:
@djc 

If you have a prelude that then implicitly imports `Debug`, this will then fail:

```rust
// The prelude with Debug:
mod prelude {
    pub(super) use core::fmt::Debug;
}

// Prelude import
use prelude::*;

use core::fmt;
use core::fmt::Display;

struct MyStruct(u32);

impl Display for MyStruct {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f) // Just delegate to the inner impl
    }
}
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=27e29cb5c637e84ea688f5ba632a17df))